internal/handler/job/events/events.go

```diff
@@ -64,81 +76,107 @@ func AddSubscriber(ctx context.Context, filter FilterParams, tags []string) (<-c
+		close(l.ch)
 	}
-	log.Info().Int("count", count).Msg("Subscriber shutdown complete")
+	log.Info().Msg("Subscriber shutdown complete")
 }
 
 // SubscriberCount counts the total number of subscribers across all topics.
 func SubscriberCount() int {
-	count := 0
-	for _, topic := range e.Topics() {
-		count += len(e.Listeners(topic))
-	}
+	mu.RLock()
+	count := len(subscribers)
+	mu.RUnlock()
 	return count
 }
 
-// PublishEvent publishes a new event. It returns a channel which can be used
-// to wait for the delivery of the event to all listeners.
-func PublishEvent(ctx context.Context, event *JobEvent) chan struct{} {
-	log := logging.LoggerFromCtx(ctx)
-	log.Debug().
-		Str("jobID", event.Job.ID).
-		Str("action", string(event.Action)).Msg("Publishing event")
-	return e.Emit(event.Job.ID, event)
-}
+// PublishEvent publishes a new event. This is a synchronous operation.
+func PublishEvent(ctx context.Context, event JobEvent) {
+	log := logging.LoggerFromCtx(ctx).With().Str("jobID", event.Job.ID).Str("action", string(event.Action)).Logger()
+	log.Debug().Msg("Publishing event to subscribers")
 
-func (filter FilterParams) createEventFilter(subscriberID string, tags []string) func(*emitter.Event) {
-	// special case: no filters means "catch-all"
-	if len(filter.JobIDs) == 0 && len(filter.ClientIDs) == 0 && len(filter.Workflows) == 0 {
-		return nil
-	}
+	mu.Lock()
+	defer mu.Unlock()
 
-	// build hash sets for faster lookup
-	jobIDSet := make(map[string]any, len(filter.JobIDs))
-	for _, s := range filter.JobIDs {
-		jobIDSet[s] = nil
-	}
-	clientIDSet := make(map[string]any, len(filter.ClientIDs))
-	for _, s := range filter.ClientIDs {
-		clientIDSet[s] = nil
-	}
-	workflowSet := make(map[string]any, len(filter.Workflows))
-	for _, s := range filter.Workflows {
-		workflowSet[s] = nil
-	}
-	return func(ev *emitter.Event) {
-		event := ev.Args[0].(*JobEvent)
-		job := event.Job
-		// check if we shall notify the client about the event
-		_, interested := jobIDSet[job.ID]
+	// the subscribers that are still active and we'll keep
+	newSubscribers := make([]Subscriber, 0, len(subscribers))
+
+	for _, l := range subscribers {
+		ctxLog := log.With().Str("subscriberID", l.subscriberID).Logger()
+
+		// check if we shall notify the subscriber about the event
+		// special case: no filters means "catch-all"
+		isCatchAll := len(l.jobIDSet) == 0 && len(l.clientIDSet) == 0 && len(l.workflowSet) == 0
+		interested := isCatchAll ||
+			mapContains(l.jobIDSet, event.Job.ID) ||
+			mapContains(l.clientIDSet, event.Job.ClientID) ||
+			(event.Job.Workflow != nil && mapContains(l.workflowSet, event.Job.Workflow.Name))
 		if !interested {
-			_, interested = clientIDSet[job.ClientID]
-		}
-		if !interested && job.Workflow != nil {
-			_, interested = workflowSet[job.Workflow.Name]
+			// keep subscriber, potentially still alive
+			newSubscribers = append(newSubscribers, l)
+			ctxLog.Debug().Msg("Subscriber not interested, skipping event notification")
```

Comment:
Hm, shouldn't we print out more information, e.g., *which* subscriber is not interested? Else we can just skip this log entry, can't we?